tests/functional/u/use/used_before_assignment_issue4761.py

```diff
@@ -10,3 +10,169 @@ def function():
+        raise error  # [used-before-assignment]
+
+
+def invalid_no_inner_else():
+    """No inner else statement."""
+    for _ in range(3):
+        try:
+            function()
+        except ValueError as verr:
+            error = verr
+        print("retrying...")
+        if function():
+            break
+    else:
+        raise error  # [used-before-assignment]
+
+
+def invalid_wrong_break_location():
+    """The break is in the wrong location."""
+    for _ in range(3):
+        try:
+            function()
+            break
+        except ValueError as verr:
+            error = verr
+            print("I give up")
+    else:
+        raise error  # [used-before-assignment]
+
+
+def invalid_no_break():
+    """No break."""
+    for _ in range(3):
+        try:
+            function()
+        except ValueError as verr:
+            error = verr
+        else:
+            pass
+    else:  # pylint: disable=useless-else-on-loop
+        raise error  # [used-before-assignment]
+
+
+def invalid_other_non_break_exit_from_loop_besides_except_handler():
+    """The continue creates another exit branch."""
+    while function():
+        if function():
+            continue
+        try:
+            pass
+        except ValueError as verr:
+            error = verr
+        else:
+            break
+    else:
+        raise error  # [used-before-assignment]
+
+
+def valid_continue_does_not_matter():
+    """This continue doesn't matter: still just one exit branch."""
+    while function():
+        try:
+            for _ in range(3):
+                if function():
+                    continue
+                print(1 / 0)
+        except ZeroDivisionError as zde:
+            error = zde
+        else:
+            break
+    else:
+        raise error
+
+
+def invalid_conditional_continue_after_break():
+    """The continue is another exit branch"""
+    while function():
+        try:
+            if function():
+                break
+            if not function():
+                continue
+        except ValueError as verr:
+            error = verr
+        else:
+            break
+    else:
+        raise error  # [used-before-assignment]
+
+
+def invalid_unrelated_loops():
+    """The loop else in question is not related to the try/except/else."""
+    for _ in range(3):
+        try:
+            function()
+        except ValueError as verr:
+            error = verr
+        else:
+            break
+    while function():
+        print('The time is:')
+        break
+    else:
+        raise error  # [used-before-assignment]
+
+
+def valid_nested_loops():
+    """The name `error` is still available in a nested else."""
+    for _ in range(3):
+        try:
+            function()
+        except ValueError as verr:
+            error = verr
+        else:
+            break
+    else:
+        while function():
+            print('The time is:')
+            break
+        else:
+            raise error
```

Comment:
I'm pretty sure we're going to still be surprised by the code that exists in the wild, but I'm also pretty confident we won't be surprised often. Those tests are thorough !